### PR TITLE
Add common error tests

### PR DESCRIPTION
**Pull request by andrew-gardener**
- make some integer error messages more consistent
- make object type error more consistent (don't use `className()`)
- Session's user should be Person (not agent)
- Event setId now requires that it is passed a string (consistent with Entity)
- Remove `entities\foaf\Agent` and replace places requiring it with `entities\agent\Agent` (easier to keep track of)
- Remove Generatable and replace places requiring it with Entity (follows spec more closely)
- Remove Targetable and replace places requiring it with Entity (follows spec more closely)
- Remove Referrable and replace places requiring it with Entity (follows spec more closely)
- Fix some events to properly require the correct generated, referrer, and target entities (AnnotationEvent, AssessmentEvent, AssessmentItemEvent, FeedbackEvent, MediaEvent, etc)

### CODE REVIEW

**Comment by andrew-gardener** [comment]
@lsloan I've started working on the common error fixtures but I'd like your input before continuing (since the tests are a fair bit different then other tests). 

So far I've done a few Entity (Entity and AggregateMeasure) and a few Event tests as examples. Please only look at https://github.com/IMSGlobal/caliper-php/pull/346/commits/9f48a6301ba420ee21eb5fdca50d353da9be6e61 since I've based the branch off the #340 PR (since I modified the tests to allow relative urls) and will rebase of develop once that's merged

**Comment by lsloan** [comment]
@andrew-gardener, I'll give an initial review later today.

**Comment by andrew-gardener** [comment]
Ended up finishing it. there are a few refactors on the larger side.

**Comment by bracken** [comment]
This looks good to me, but I'm not set up to run these tests locally and our build is running here. So if you are +1 and they run locally for you I'm happy to merge.

**Comment by lsloan** [comment]
My apologies for saying I would review this and never getting back to it.  My UMich obligations have changed considerably.  I miss having time to work on this project.

I saw @bracken's message today and it prompted/guilted me to finally start reviewing it.  I have no objections to merging this as long as he's satisfied with it.

I was wondering, though, why the classes like `Targetable`, `Generatable`, etc. were removed.  Is it that the specification says any kind of entity can be used in those properties that had required more specific ones before?  (I don't mean to open a debate about it.  I'll be satisfied with the answer "RTFM" if that's what you'd like to answer.)

**Comment by andrew-gardener** [comment]
I'll try and get the github actions finished first then. That way the tests can run in that. Just doing a few updates to get the tests to run on php 8 (7.3 - 8)

> I was wondering, though, why the classes like Targetable, Generatable, etc. were removed. Is it that the specification says any kind of entity can be used in those properties that had required more specific ones before? (I don't mean to open a debate about it. I'll be satisfied with the answer "RTFM" if that's what you'd like to answer.)

I'm not 100% sure why they were there originally. I removed them because they were becoming redundant since basically every entity is `Targetable` and `Generatable` due to new profiles.

**Comment by andrew-gardener** [comment]
I can try to add them back in if desired.

**Comment by bracken** [comment]
@lsloan don't worry about it! My review was rather cursory and just to kind of make sure there isn't anything obvious. All these kinds of things just move forward in little spurts as people have need and time to move them forward.

**Comment by andrew-gardener** [comment]
Please see #348 for github action automated tests

**Comment by lsloan** [comment]
> > I was wondering, though, why the classes like Targetable, Generatable, etc. were removed…
> 
> I'm not 100% sure why they were there originally. I removed them because they were becoming redundant since basically every entity is `Targetable` and `Generatable` due to new profiles.

When I worked on caliper-php, I based it on caliper-java, which was mostly complete for the version of the specification at that time.  There were only certain entity types that could be used as a `target` property or a `generated` property, etc.  For example, there was never any example of an event that would result in the generation of a `Person` entity, so that entity wouldn't implement `Generatable`. The `generated` property would only allow entities that had implemented that interface and an error could be given for any entities that didn't.  It was meant to prevent the output of Caliper events that don't confirm to the specification.

These interface classes didn't do much (or even anything in most cases), but they identified groups of entities that could be used for specific properties.  They're just used as "marker classes" and they eliminated the need to write logic like `if $entity not type ClassA and not type ClassB and not… throw Exception("entity not of types ClassA, ClassB, …")`.

Since I don't know what the latest version of the spec. allows, if it says any kind of entity can be used in any property (or if not explicitly, it pretty much works out that way), then I understand why these marker classes are no longer needed.  If that's not the case, it seems like there needs to be some other mechanism to check whether the entities being used are of the allowed types.

PHP continues to change and it may have better ways to do this kind of thing.  Rather than using interfaces as marker classes, maybe it's better to use "traits".  I don't know.  I've not used PHP as much in the past 1–2 years.

To go off on a tangent slightly:  I remember someone at IMS, it may have been Colin or Mark Leuba, didn't like that caliper-php threw so many exceptions because it gave the appearance that the package is buggy.  I disagreed.  Throwing exceptions is a legitimate coding technique.  It prevents the user from doing something wrong, and puts the responsibility for fixing it back in their hands.  I think strict type checking performs a similar role.  Either way, we're trying to ensure that the Caliper JSON that comes out of this package conforms to the standard.

Again, I apologize for not reviewing this months ago and bringing this up earlier.  I don't want to seem like the bad guy coming in at the last moment and complaining about everything.  But I thought I should ask about it just to make sure these changes are what the specification intends.

**Comment by lsloan** [comment]
Skimming the Caliper 1.2 specification I see that no events have section headings like "supported target entities" or "supported generated entities".  The 1.1 specification had those sections for event types as needed.  Maybe that's the indication that any entity type can be used anywhere.

**Comment by andrew-gardener** [comment]
Alright, that explains the origin a lot more. Right now I believe most of that kind of logic is now within the events objects themself (with occasional conditional around specific profile used). Or at least that's what I've been doing when making changes. I think it makes it easier to follow the specific specs for the events (since you just overload the function with what the spec allow in that event) but I'm open to changes. Again, I only removed them since it seemed redundant at the time and I was cleaning up a bunch of other little things that shifted off spec over time.

There is probably room for a good old refactor to use some more modern php to solve some of this. I don't primary program in php though so I'd need to refresh myself to do that

(also I'm also in favor of throwing more errors if people put in the incorrect values. better to try and get people to follow the spec than not. The removal of `Targetable` and `Generatable`, etc I think helps with this a bit since its more explicit in what entities are expected in the errors).

**Comment by andrew-gardener** [comment]
@bracken sorry took a while to rebase. Had to rename all the common error files (replace `-` with `_`) to stop a warnings during test runs

Also looks like the tests are running properly on a PR https://github.com/IMSGlobal/caliper-php/actions/runs/579705483

**Comment by lsloan** [comment]
> I think it makes it easier to follow the specific specs for the events (since you just overload the function with what the spec allow in that event) but I'm open to changes. Again, I only removed them since it seemed redundant at the time and I was cleaning up a bunch of other little things that shifted off spec over time.
> …
> (also I'm also in favor of throwing more errors if people put in the incorrect values. better to try and get people to follow the spec than not. The removal of `Targetable` and `Generatable`, etc I think helps with this a bit since its more explicit in what entities are expected in the errors).

Yes, that makes sense.  The more I think about it and look at the spec. again, I feel better about your approach.  My earlier statement that the 1.2 spec seems to allow any kind of entity type is wrong.  I see now that the spec no longer uses the phrase "supported generated entities".  Instead, each event has a table of generated entity types allowed based on other criteria in the event.

A benefit of `Targetable`, `Generatable`, etc. is that violations are caught by the interpreter or other steps earlier in the process.  Catching violations in the code is still needed, of course.  In an ideal world, we'd do both, but that's more difficult to maintain.  Also in an ideal world, we could say that the property of a subclass must be one of a union of types specific to the class, not a non-specific collection, like `Generatable`.

PS: Nice job with the GH actions feature.

**Review by lsloan — APPROVED** [review]
I have some suggestions, but they are not major.  They're certainly not important enough to hold back the merge, so I'll approve this PR.

I focused mainly on reviewing changes in the `src` directory.  Most of the tests are specific to the contents of the fixtures, which I'm not very familiar with now.  I noticed a minor problem with quoting style in the tests, but that's all.

A lot of the implementation is very repetitive.  I notice this especially in the new code.  Looking at all the code together, it's tough to distinguish which parts are new or old because they look the same.  It would be nice if the implementation could use more generalized logic and a JSON schema of the Caliper specification to reduce the code.

File: src/entities/Entity.php [hunk 1]
```diff
@@ -7,7 +7,7 @@
 use IMSGlobal\Caliper\util;
 use IMSGlobal\Caliper\util\ClassUtil;
 
-class Entity extends ClassUtil implements \JsonSerializable, entities\schemadotorg\Thing {
+class Entity extends ClassUtil implements \JsonSerializable, entities\schemadotorg\Thing, Referrable {
```
[the remark this answers is not in the document]
Comment (andrew-gardener) [reply src/entities/Entity.php]: Removed `Referrable` interface (was only used for Entity)

File: src/entities/session/LtiSession.php [hunk 2]
```diff
@@ -5,7 +5,7 @@
 use IMSGlobal\Caliper\entities;
 
 class LtiSession extends Session {
-    /** @var mixed */
+    /** @var array (associative) */
```
[the remark this answers is not in the document]
Comment (andrew-gardener) [reply src/entities/session/LtiSession.php]: Changed to array (the `(associative)` was more of a comment. The spec has it as an `Object` and I figured the the most honest way would be to reference it as an associative array. Honestly not used to using comments to declare type in php so not sure if is valid).

File: tests/commonErrorTests/Event_MalformedSessionNotAStringTest.php [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+                    ->setOrganization(
+                        (new Organization('https://example.edu/terms/201601/courses/7/sections/1'))->makeReference())
+                    ->setRoles(
+                        [new Role(Role::LEARNER)])
+                    ->setStatus(
+                        new Status(Status::ACTIVE))
+                    ->setDateCreated(
+                        new \DateTime('2016-08-01T06:00:00.000Z'))
+            )
+            ->setSession("This is not an object or an IRI");
```
[the remark this answers is not in the document]
Comment (andrew-gardener) [reply tests/commonErrorTests/Event_MalformedSessionNotAStringTest.php]: Fixed

**Comment by andrew-gardener** [comment]
Also fyi I do not have permissions to merge the PR so someone else needs to do it if its all good!

**Comment by lsloan** [comment]
OK.  No problem.

I was surprised you used force push for the quotes correction, but that's cool.  I just don't use force push very often.

**Comment by andrew-gardener** [comment]
I generally try to work in a one commit feature branch workflow (with rebasing+amending). I think its common in open source project (on the larger side) so that it easier to track down problems in the history (only one commit with a useful message explaining the full changes). Its been working pretty well in a lot of projects I've been working on

**Comment by bracken** [comment]
that feature branch with rebase/amending is also my preferred workflow andrew. Thanks for pushing all these through!

**Comment by lsloan** [comment]
Thanks for the info about the one commit feature branch workflow, @andrew-gardener and @bracken.  I'll keep it in mind.

**Comment by andrew-gardener** [comment]
> Does that workflow have much benefit over making multiple commits, then doing a squash merge at the end?
Not really but it saves the potential issue where it isn't squashed+rebased at the end (due to accidentally forgetting/changing commit settings or the repo not being setup for force squash+rebase for PRs).

Otherwise it makes it easier if you want to manually rebase (since you only need to replay the one commit and deal with merge conflicts once) and makes the PR message easier to maintain and update (since you can c+p the commit message instead of tracking the changes in the commit)

**Pull request merged**